crates/oxc_linter/src/rules/promise/no_nesting.rs

```diff
@@ -147,51 +147,59 @@ fn can_safely_unnest(
     closest: &CallExpression,
     ctx: &LintContext,
 ) -> bool {
-    let mut closest_cb_scope_bindings = vec![];
+    let mut safe_to_unnest: bool = true;
 
     closest.arguments.iter().for_each(|new_expr| {
-        let Some(arg_expr) = new_expr.as_expression() else {
-            return;
-        };
-
-        match arg_expr {
-            Expression::ArrowFunctionExpression(arrow_expr) => {
-                let func_scope = arrow_expr.scope_id();
-                closest_cb_scope_bindings =
-                    ctx.scopes().get_bindings(func_scope).iter().collect_vec();
-            }
-            Expression::FunctionExpression(func_expr) => {
-                let func_scope = func_expr.scope_id();
-                closest_cb_scope_bindings =
-                    ctx.scopes().get_bindings(func_scope).iter().collect_vec();
+        if let Some(arg_expr) = new_expr.as_expression() {
+            match arg_expr {
+                Expression::ArrowFunctionExpression(arrow_expr) => {
+                    let scope = arrow_expr.scope_id();
+                    if usage_of_closest_cb_vars(scope, call_expr, ctx) {
+                        safe_to_unnest = false;
+                    }
+                }
+                Expression::FunctionExpression(func_expr) => {
+                    let scope = func_expr.scope_id();
+                    if usage_of_closest_cb_vars(scope, call_expr, ctx) {
+                        safe_to_unnest = false;
+                    }
+                }
+                _ => {}
             }
-            _ => {}
-        }
+        };
     });
 
-    if let Some(cb_span) = call_expr.arguments.first().map(GetSpan::span) {
-        // Now check for references in cb_span to variables defined in the closest parent cb scope.
-        // In the given example we would loop through all bindings in the closest
-        // parent scope a,b,c,d.
-        //
-        //  .then((a,b,c) => {
-        //    const d = 5;
-        //    getB(a).then(d => getC(a, b)) });
-        //                // ^^^^^^^^^^^^^^ <- `cb_span`
-        for (_, binding_symbol_id) in closest_cb_scope_bindings {
+    safe_to_unnest
+}
+
+/// Check check for references in cb_span to variables defined in the closest parent cb scope.
+/// In the given example we would loop through all bindings in the closest
+/// parent scope a,b,c,d.
+///
+///  .then((a,b,c) => { // closest_cb_scope_id
+///    const d = 5;
+///    getB(a).then(d => getC(a, b)) });
+///                // ^^^^^^^^^^^^^^ <- `cb_span`
+fn usage_of_closest_cb_vars(
+    closest_cb_scope_id: ScopeId,
+    cb_call_expr: &CallExpression,
+    ctx: &LintContext,
+) -> bool {
+    if let Some(cb_span) = cb_call_expr.arguments.first().map(GetSpan::span) {
```

Comment:
can you move this `if let` into `can_safely_unnest` and rather than passing `cb_call_expr` just pass either the argument or the span? that way we can avoid doing the iter work above if there is no first arg

Reply:
Good idea! That's it done :)